### ISSUE TITLE
Buildozer spec

### ISSUE THREAD

**Issue by ChristopherYoon89** [Blackysh/ffmpeg-kit-python#2]
Hello Arqies!

Thanks for your kindness and your offer to help. 

This is my buildozer spec file:

```
[app]

# (str) Title of your application
title = myApp

# (str) Package name
package.name = myapp

# (str) Package domain (needed for android/ios packaging)
package.domain = org.test

# (str) Source code where the main.py live
source.dir = .

# (list) Source files to include (let empty to include all the files)
source.include_exts = py,png,jpg,kv,atlas,ini,json

# (list) Include folders 
source.include_patterns = user/*

# (list) Source files to exclude (let empty to not exclude anything)
#source.exclude_exts = spec

# (list) List of directory to exclude (let empty to not exclude anything)

# (list) List of exclusions using pattern matching
#source.exclude_patterns = license,images/*/*.jpg

# (str) Application versioning (method 1)
#version.regex = __version__ = ['"](.*)['"]
#version.filename = %(source.dir)s/main.py

# (str) Application versioning (method 2)
version = 0.1

# (list) Application requirements
requirements = python3,kivy==2.2.1,kivymd==1.1.1,android,plyer,Wave,pydub,urllib3,kivy-garden.xcamera,kivy-garden.zbarcam,Pillow,pyzbar,jnius

# (str) Custom source folders for requirements
# Sets custom source for any requirements with recipes
# requirements.source.kivy = ../../kivy

# (list) Garden requirements
#garden_requirements =

# (str) Presplash of the application
presplash.filename = %(source.dir)s/data/presplash.png

# (str) Icon of the application
icon.filename = %(source.dir)s/data/icon.png

# (str) Supported orientation (one of landscape, portrait or all)
orientation = portrait

# (bool) Indicate if the application should be fullscreen or not
fullscreen = 0


#
# Android specific
#

# (list) Permissions
android.permissions = RECORD_AUDIO,WAKE_LOCK,CAMERA,INTERNET

# (int) Android API to use
android.api = 33

# (int) Minimum API required (8 = Android 2.2 devices)
android.minapi = 24

# (int) Android SDK version to use
android.sdk = 33

# (str) Android NDK version to use
#android.ndk = 9c

# (bool) Use --private data storage (True) or --dir public storage (False)
#android.private_storage = True

# (str) Android NDK directory (if empty, it will be automatically downloaded.)
#android.ndk_path =

# (str) Android SDK directory (if empty, it will be automatically downloaded.)
#android.sdk_path =

# (str) python-for-android git clone directory (if empty, it will be automatically cloned from github)
#android.p4a_dir =

# (list) python-for-android whitelist
#android.p4a_whitelist =

# (str) Android entry point, default is ok for Kivy-based app
#android.entrypoint = org.renpy.android.PythonActivity

# (list) List of Java .jar files to add to the libs so that pyjnius can access
# their classes. Don't add jars that you do not need, since extra jars can slow
# down the build process. Allows wildcards matching, for example:
# OUYA-ODK/libs/*.jar

android.add_jars = smart-exception-common-0.2.1.jar,smart-exception-java-0.2.1.jar

android.add_aars = ffmpeg-kit-release.aar

# (list) List of Java files to add to the android project (can be java or a
# directory containing the files)
#android.add_src =

# (str) python-for-android branch to use, if not master, useful to try
# not yet merged features.
#android.branch = master

# (str) OUYA Console category. Should be one of GAME or APP
# If you leave this blank, OUYA support will not be enabled
#android.ouya.category = GAME

# (str) Filename of OUYA Console icon. It must be a 732x412 png image.
#android.ouya.icon.filename = %(source.dir)s/data/ouya_icon.png

# (str) XML file to include as an intent filters in <activity> tag
#android.manifest.intent_filters =

# (list) Android additionnal libraries to copy into libs/armeabi
#android.add_libs_armeabi = libs/android/*.so
#android.add_libs_armeabi_v7a = libs/android-v7/*.so
#android.add_libs_x86 = libs/android-x86/*.so
#android.add_libs_mips = libs/android-mips/*.so

# (bool) Indicate whether the screen should stay on
# Don't forget to add the WAKE_LOCK permission if you set this to True
#android.wakelock = False

# (list) Android application meta-data to set (key=value format)
#android.meta_data =

# (list) Android library project to add (will be added in the
# project.properties automatically.)
#android.library_references =

#
# iOS specific
#

# (str) Name of the certificate to use for signing the debug version
# Get a list of available identities: buildozer ios list_identities
#ios.codesign.debug = "iPhone Developer: <lastname> <firstname> (<hexstring>)"

# (str) Name of the certificate to use for signing the release version
#ios.codesign.release = %(ios.codesign.debug)s


[buildozer]

# (int) Log level (0 = error only, 1 = info, 2 = debug (with command output))
log_level = 2

# (int) Display warning if buildozer is run as root (0 = False, 1 = True)
warn_on_root = 1


#    -----------------------------------------------------------------------------
#    List as sections
#
#    You can define all the "list" as [section:key].
#    Each line will be considered as a option to the list.
#    Let's take [app] / source.exclude_patterns.
#    Instead of doing:
#
#[app]
#source.exclude_patterns = license,data/audio/*.wav,data/images/original/*
#
#    This can be translated into:
#
#[app:source.exclude_patterns]
#license
#data/audio/*.wav
#data/images/original/*
#


#    -----------------------------------------------------------------------------
#    Profiles
#
#    You can extend section / key with a profile
#    For example, you want to deploy a demo version of your application without
#    HD content. You could first change the title to add "(demo)" in the name
#    and extend the excluded directories to remove the HD content.
#
#[app@demo]
#title = My Application (demo)
#
#[app:source.exclude_patterns@demo]
#images/hd/*
#
#    Then, invoke the command line with the "demo" profile:
#
#buildozer --profile demo android debug
```

Thanks again for your help!

**Comment by arqies** [comment]
There are few things that may have caused the issue (I am not certain, and can't test rn) but try:

- Add `pyjnius` alongside/replacing jnius in the requirements (Its a mistake in the video, fixed in the written documentation)
- `android.minapi = 24` -> `25` (24 doesn't seem to work for me)
- make sure you are using Java 8 instead of newer Java versions.

If issue persists or different issue arises, let me know.

**Comment by ChristopherYoon89** [comment]
Thanks, I got it run. I changed the minapi to 25 and added pyjnius. Now, ffmpeg is running on my phone. But I had to execute the command `buildozer clean app` first and build it again, then it worked. Thanks again for your help.

**Comment by ChristopherYoon89** [comment]
Just one more thing. 

Can you give me a high-level explanation of how it works to add third party libraries to python kivy on Android? Do I have to config cradle or add the libraries to the ndk or how does your ffmpeg-kid work? 

I am asking because I have built my own binaries because of the large file size of the ffmpeg library and I don't need a lot of functionalities such as video etc. Would appreciate if you can just give a short overview or even a detailed description.

**Comment by arqies** [comment]
Can you give me a high-level explanation of how it works to add third party libraries to python kivy on Android? You can add Java libraries using [pyjnius](https://pyjnius.readthedocs.io/). But there will be errors, related to permissions or smth else, you will need to modify them to make them work, which is what I did in this fork. I  rebuilt the java ffmpegkit with those errors fixed, and the orignal one required code that couldn't be written in Python, so I also had to add Java function to  the orignal, that could work with features of python. 

How does this FFmpeg kit work? This fork is just a modification of arthentica's Java FFmpeg kit. And honestly, I haven't even read the orginal Java code for anything other than error-fixing and creating functions within Java for python use. All I know is how the Java libarary communicates with python, not what it does in Java and what about ffmpeg.

And what binaries have you built, modified ffmpeg? , or some other Java libararies. If its neither of these I can't help, but if it is one of these, please provide further information about what it is, etc.

**Comment by ChristopherYoon89** [comment]
Thanks for your response. Yes, I originally built my own ffmpeg binaries. But I found a more elegant solution to my problem, in which I don't rely on ffmpeg anymore. I built my own Audio Recorder directly accessing the Android hardware and its constants and generate the audio file using the built-in output formats and encoders instead of relying on outdated Python libraries such as plyer. So, I don't require the ffmpeg library to convert the audio file because their native output formats and encoders include the one that I need.
